external-crates/move/documentation/book/src/enums.md

```diff
@@ -0,0 +1,335 @@
+        VariantWithPositionalFields(u64, bool),
+    }
+    public enum Baz has drop {
+        VariantWithNamedFields { x: u64, y: bool, z: Bar },
+    }
+}
+```
+
+Enums cannot be recursive in any of their variants, so the following definitions of an enum are not
+allowed since they would be recursive in at least one variant:
+
+```move
+module a::m {
+    public enum Foo {
+        Recursive(Foo),
+        //        ^ error: recursive enum variant
+    }
+    public enum List {
+        Nil,
+        Cons { head: u64, tail: List },
+        //                      ^ error: recursive enum variant
+    }
+    public enum BTree<T> {
+        Leaf(T),
+        Node { left: BTree<T>, right: BTree<T> },
+        //           ^ error: recursive enum variant
+    }
+}
+```
+
+## Visibility
+
+All enums are declared as `public`. This means that the type of the enum can be referred to from any
+other module. However, the variants of the enum, the fields within each variant, and the ability to
+create or destroy variants of the enum are internal to the module that defines the enum.
+
+In the future, we plan on adding the ability to declare enums to have public variants, but this is
+currently not supported.
+
+### Abilities
+
+Just like with structs, by default an enum declaration is linear and ephemeral. In order to use an
+enum value in a non-linear or non-ephemeral way -- i.e., copied, dropped, or stored in an
+[object](./abilities/object.md) -- you need to grant it additional [abilities](./abilities.md) by
+annotating them with `has <ability>`:
+
+```move
+module a::m {
+    public enum Foo has copy, drop {
+        VariantWithNoFields,
+    }
+}
+```
+
+The ability declaration can occur either before or after the enum's variants, however only one or
+the other can be used, and not both. If declared after the variants, the ability declaration must be
+terminated with a semicolon:
+
+```move
+module a::m {
+    public enum PreNamedAbilities has copy, drop { Variant }
+    public enum PostNamedAbilities { Variant } has copy, drop;
+    public enum PostNamedAbilitiesInvalid { Variant } has copy, drop
+    //                                                              ^ ERROR! missing semicolon
+
+    public enum NamedInvalidAbilities has copy { Variant } has drop;
+    //                                                     ^ ERROR! duplicate ability declaration
+}
+```
+
+For more details, see the section on
+[annotating abilities](./abilities.md#annotating-structs-and-enums).
+
+## Naming
+
+Enums and variants within enums must start with a capital letter `A` to `Z`. After the first letter,
+enum names can contain underscores `_`, letters `a` to `z`, letters `A` to `Z`, or digits `0` to
+`9`.
+
+```move
+public enum Foo { Variant }
+public enum BAR { Variant }
+public enum B_a_z_4_2 { V_a_riant_0 }
+```
+
+This naming restriction of starting with `A` to `Z` is in place to give room for future language
+features. It may or may not be removed later.
```

Comment:
```suggestion
features.
```
This may or may not be useful info. If you deem important, I'd suggest just "It might be removed later."